### ISSUE TITLE
Leverage image preloading in kubemark tests

### ISSUE THREAD

**Issue by tosi3k** [kubernetes/kubernetes#90833]
**What would you like to be added**:
The change in #90806 is temporary - instead of just adding an annotation to every kubemark node to simply increase node objects sizes, we'd like to leverage image preloading (from clusterloader in perf-tests repo) once it supports kubemark clusters to better reflect the structure of node objects in kubemark tests. 

**Why is this needed**:
Explained in previous paragraph.

**Comment by irbekrm** [comment]
Hi! Could I please pick this up as my first issue?

**Comment by mm4tt** [comment]
Hey @irbekrm, that's great to hear!

The extremely high level explanation of what we'd like to do is in this comment - https://github.com/kubernetes/kubernetes/pull/90722#discussion_r420045355

I'll try to post more detailed steps here soon, but a good start would be to familiarize yourself with the kubemark code, especially the part where we mock CRI (I think this is the entry point - https://github.com/kubernetes/kubernetes/blob/8d54a84507c622e1552373d5d125c84ee22d1465/cmd/kubemark/hollow-node.go#L217)  to see whether it would be possible to have hollow_node take images from the real nodes

/cc @wojtek-t

**Comment by irbekrm** [comment]
Thanks for your advice @mm4tt ! I read through the kubemark code and was then trying to understand what the change in #90806 actually does. I am having a bit of hard time finding where in the codebase the annotation that is overwritten in #90806 is actually used, was wondering if you could point me in the right direction please?

**Comment by mm4tt** [comment]
Hey @irbekrm, the change in https://github.com/kubernetes/kubernetes/pull/90806 is actually a workaround and it has nothing in common with the long-term solution we'd like to have. What we did in https://github.com/kubernetes/kubernetes/pull/90806 was to simply add a large label to kubemark node object to artificially increase their size. Long term, we'd like to have kubemark nodes mimic the structure on the real nodes. The most important part of it is to have the kubemark nodes take images from the real nodes (i.e. the nodes where the kubemark's hollow_node pods are running). That's why I pointed you to the place where we pass fake image service to the hollow kubelet. If we could implement and pass there a different implementation of the image service that would be still fake but would return the image list from the real node then it would achieve what we need. 

Later part is to use the clusterloader2 feature called imagepreload (https://github.com/kubernetes/perf-tests/tree/master/clusterloader2/pkg/imagepreload) to load more images into the real nodes to make our test clusters more similar to the prod clusters.
It will require some changes in Cl2, to make it work properly in kubemark clusters, but it should be relatively simple.

**Comment by wojtek-t** [comment]
FTR: One place where the current mechanism is causing significant differences comparing to reality is for components that list/watch only ObjectMetadata (as=PartialObjectMetadata(List)). addon-resizer being one example here. IIRC in-tree garbage-collector is another one.
@mm4tt ^^

**Comment by irbekrm** [comment]
Unfortunately I think it is going to take me too long to figure out cluster loader and kubemark. I will re-assign myself from this issue to give someone else a chance.

**Comment by mm4tt** [comment]
Thanks, @irbekrm. And no worries, I'm aware that the issue is complicated. I need to add more details here, but I'm quite overloaded recently and I'm not sure when I'll have time to do it. 

@wojtek-t, thanks for providing these examples. It's clear to me that the temporary solution we have in place is not ideal, but it's better than nothing. I hope that we'll implement the full solution eventually.

**Comment by alam0rt** [comment]
Hey, I might have a go of this one.

**Comment by alam0rt** [comment]
@mm4tt, i understand once I create a PR we can run some tests, but before implement anything, is there an easy way to test my changes to the fake image service?

**Comment by mm4tt** [comment]
I'm not sure whether we have any unit tests for kubemark, but you could add some if you'd like. Other than that, nothing else really comes to my mind other than just seeing if kubemark builds with your change locally and later running real test via presubmits.

**Comment by alam0rt** [comment]
Alright, I'll slap something together and we can see what holds.

**Comment by artback** [comment]
Active @alam0rt ?

**Comment by alam0rt** [comment]
Feel free to pick it up. Have been quite busy.

**Comment by alam0rt** [comment]
I have a bit more free time now. Will attempt again.

**Comment by Dinesh-Balakrishnan** [comment]
Hi! I would like to try to start contributing to Kubernetes. I'll attempt to work on this issue.

**Comment by lyzs90** [comment]
@alam0rt @Dinesh-Balakrishnan are you guys still on this?

> If we could implement and pass there a different implementation of the image service that would be still fake but would return the image list from the real node then it would achieve what we need.

@mm4tt i'm wondering if is this the right approach for part 1?
- instead of passing a fake imageService and runtimeService to the [HollowKubelet](https://github.com/kubernetes/kubernetes/blob/8d54a84507c622e1552373d5d125c84ee22d1465/cmd/kubemark/hollow-node.go#L212), pass in the real `RemoteImageService` and `RemoteRuntimeService` set up with the dockershim endpoint
- set HollowKubelet `ContainerRuntime` flag to docker
- set HollowKubelet `RemoteRuntimeEndpoint` and `RemoteImageEndpoint` config to dockershim

the last two points are because I got a little confused when i reached https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/kubelet.go#L317 where these two services gets overridden. i couldn't see how the mocks were going to be used past that point

**Comment by lyzs90** [comment]
I missed the part where hollow nodes are run as pods, so we'll need to modify the hollow node template to bind mount the docker (or cri) socket & binary? I think you'll need this for image preloading later

Let me know if this is what you had in mind and I can put in the necessary changes

**Comment by wojtek-t** [comment]
@lyzs90 - sorry for delayed response - vacation period

What you wrote above makes sense, with the differences:
1. that we're deprecating and removing dockershim. I think our tests are already using containerd, so you need to contact the containerd (CRI?) sockets or something like that.
2. I think we don't want to touch RemoteRuntimeService - we only want to touch ImageEndpoint. The fact that we have fake runtime is actually by design and expected.

**Comment by lyzs90** [comment]
@wojtek-t no worries, welcome back! 

thanks for clarifying. as for local testing, would it suffice to start a hollow node pod and see if it can pull an image?

**Comment by wojtek-t** [comment]
> thanks for clarifying. as for local testing, would it suffice to start a hollow node pod and see if it can pull an image?

It should be. And you will also see what happens as part of presubmit (we have mandatory kubemark presubmit in perf-tests repo and option in k/k)

**Comment by mm4tt** [comment]
Thanks @lyzs90! Your approach looks good, let's see how it works out in the kubemark presubmit. 

FTR, the second part of this issue is blocked by https://github.com/kubernetes/perf-tests/issues/1619 - CL2 imagepreload currently doesn't work with containerd.

**Comment by mm4tt** [comment]
The three remaining AIs are:

1.  [ ] Stop setting the [KUBEMARK_NODE_OBJECT_SIZE_BYTES](https://github.com/kubernetes/test-infra/blob/afd239de5b925758f33ca8c4e6b8e50d5dbcb6a0/config/jobs/kubernetes/sig-scalability/sig-scalability-presets.yaml#L23) env in the kubemark tests
2. [ ] Clean up the not-needed anymore code in start-kubemark.sh - https://github.com/kubernetes/kubernetes/blob/6b1393a3c597aaf2546dafe5707ad12298fcfd09/test/kubemark/start-kubemark.sh#L228-L227
3. [ ] Start using image-preload in CL2 tests (let me think more about it).

@lyzs90, would you mind taking care of the first two clean-up AIs?

**Comment by lyzs90** [comment]
@mm4tt sure, i can take it up

**Comment by lyzs90** [comment]
> Start using image-preload in CL2 tests (let me think more about it).

@mm4tt what kind of images were you planning to load? Arbitrary ones just to pad the size?

**Comment by mm4tt** [comment]
Yeah, for the node object the only thing that counts is the number of the images. That being said, the images will be actually downloaded on each node, so we need to make sure that:

1. Images are as small as possible
2. They are hosted in gcr.io to makes sure we don't run into docker-hub (or other registry) rate-limiting issues

**Comment by lyzs90** [comment]
Ah. What are the scalability concerns with regards to large number of images?

I take it that size minimization is for practical / cost reasons?

**Comment by mm4tt** [comment]
The more images you have on the node, the bigger the node object is in etcd. As in a large cluster there is thousand of node objects and they are updated pretty frequently, it directly translates to performance of etcd / kube-apiserver.

Yeah, size minimization is for cost reasons.

**Comment by lyzs90** [comment]
How about building many scratch images with minor differences like a file with a random name? 

We could use image preload in its current implementation to mass pull. But the args on these tests will be very long and even gcr has some [quotas](https://cloud.google.com/container-registry/quotas). 

Alternatively we could extend the functionality to support downloading a tarball, then use `ctr image import` to preload the images.

**Comment by mm4tt** [comment]
I wouldn't worry about quotas. They are per client IP and more than enough to satisfy our needs - we need to download O(10-20) images per node / clientIP.

Scratch images sound like a great idea. We could create some scripts in perf-tests/util-images that will allow us to create and push N such images to gcr.

**Comment by lyzs90** [comment]
Then for the test cmd args, instead adding a long comma separated string to each test, do you think we should support globs?

**Comment by mm4tt** [comment]
I wouldn't worry about the arg being too long. Well, it won't be pretty but we can live with that as long as it works. We can also set it via an env var to keep the command line simple. Long term, I believe we should invest in a config file to set things like this.
